[FEATURE] Encryption of messages

### Description

Right now, the messages are just being sent to the database raw, which isn't right, we need a way to encrypt them and the user ids too

### Screenshots

_No response_

### Additional information

_No response_

### 👀 Have you checked if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Contributing Guidelines?

- [X] I have read and understood the rules in the [Contributing Guidelines](https://github.com/Dun-sin/Whisper/blob/main/CONTRIBUTING.md)

### Are you willing to work on this issue ?

None

I have been getting this error when trying to do `npm install` 

`npm ERR! gyp ERR! stack Error: Could not find any Visual Studio installation to use
npm ERR! gyp ERR! stack     at VisualStudioFinder.fail (C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:122:47)
npm ERR! gyp ERR! stack     at C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:75:16
npm ERR! gyp ERR! stack     at VisualStudioFinder.findVisualStudio2013 (C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:363:14)
npm ERR! gyp ERR! stack     at C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:71:14
npm ERR! gyp ERR! stack     at C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:384:16
npm ERR! gyp ERR! stack     at C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\util.js:54:7
npm ERR! gyp ERR! stack     at C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\util.js:33:16
npm ERR! gyp ERR! stack     at ChildProcess.exithandler (node:child_process:427:5)
npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
npm ERR! gyp ERR! stack     at maybeClose (node:internal/child_process:1091:16)
npm ERR! gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:302:5)
npm ERR! gyp ERR! System Windows_NT 10.0.22000`

Im using the latest version of visual studio code on my machine

> I have been getting this error when trying to do `npm install`
> 
> `npm ERR! gyp ERR! stack Error: Could not find any Visual Studio installation to use npm ERR! gyp ERR! stack at VisualStudioFinder.fail (C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:122:47) npm ERR! gyp ERR! stack at C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:75:16 npm ERR! gyp ERR! stack at VisualStudioFinder.findVisualStudio2013 (C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:363:14) npm ERR! gyp ERR! stack at C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:71:14 npm ERR! gyp ERR! stack at C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:384:16 npm ERR! gyp ERR! stack at C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\util.js:54:7 npm ERR! gyp ERR! stack at C:\Users\denni\PersonalProjects\Whisper\node_modules\node-gyp\lib\util.js:33:16 npm ERR! gyp ERR! stack at ChildProcess.exithandler (node:child_process:427:5) npm ERR! gyp ERR! stack at ChildProcess.emit (node:events:513:28) npm ERR! gyp ERR! stack at maybeClose (node:internal/child_process:1091:16) npm ERR! gyp ERR! stack at ChildProcess._handle.onexit (node:internal/child_process:302:5) npm ERR! gyp ERR! System Windows_NT 10.0.22000`
> 
> Im using the latest version of visual studio code on my machine

Where did you run this?

--- a/client/package.json
+++ b/client/package.json
@@ -22,6 +22,7 @@
 		"@rsuite/icons": "^1.0.3",
 		"axios": "^1.4.0",
 		"bad-words-next": "^2.2.1",
+		"crypto-js": "^4.1.1",
 		"emoji-picker-react": "^4.5.2",
 		"lodash": "^4.17.21",
 		"markdown-it": "^13.0.1",

--- a/client/src/components/Chat.jsx
+++ b/client/src/components/Chat.jsx
@@ -39,6 +39,7 @@ import MessageSeen from './Chat/MessageSeen';
 import MessageInput from './Chat/MessageInput';
 import DropDownOptions from './Chat/DropDownOption';
 import PreviousMessages from './Chat/PreviousMessages';
+import decryptMessage from 'src/lib/decryptMessage';
 
 
 const inactiveTimeThreshold = 180000; // 3 mins delay
@@ -59,7 +60,7 @@ const Chat = () => {
 	const [openPreviousMessages, setOpenPreviousMessages] = useState(null);
 	const [badwordChoices, setBadwordChoices] = useState({});
 
-	const { messages: state, addMessage, updateMessage, removeMessage, editText, receiveMessage } = useChat();
+	const { messages: state, addMessage, updateMessage, removeMessage, receiveMessage } = useChat();
 	const { authState, dispatchAuth } = useAuth();
 	const { logout } = useKindeAuth();
 	const socket = useContext(SocketContext);
@@ -104,22 +105,7 @@ const Chat = () => {
 		[state, app.currentChatId]
 	);
 
-	const doSend = async ({ senderId, room, tmpId = uuid(), message, time, containsBadword }) => {
-		try {
-			addMessage({
-				senderId,
-				room,
-				id: tmpId,
-				message,
-				time,
-				status: 'pending',
-				containsBadword
-			});
-		} catch {
-			logOut();
-			return false;
-		}
-
+	const doSend = async ({ senderId, room, message, time, containsBadword }) => {
 		try {
 			const sentMessage = await sendMessage({
 				senderId,
@@ -129,18 +115,28 @@ const Chat = () => {
 				containsBadword
 			});
 
+			addMessage({
+				senderId,
+				room,
+				id: sentMessage.id,
+				message,
+				time,
+				status: 'pending',
+				containsBadword
+			});
+
 			try {
-				updateMessage(tmpId, sentMessage);
+				updateMessage(sentMessage);
 			} catch {
 				logOut();
 				return false;
 			}
 		} catch (e) {
 			try {
-				updateMessage(tmpId, {
+				updateMessage({
 					senderId,
 					room,
-					id: tmpId,
+					id: uuid(),
 					message,
 					time,
 					status: 'failed',
@@ -183,18 +179,17 @@ const Chat = () => {
 
 		if (editing.isediting === true) {
 			try {
-				const oldMessage = getMessage(editing.messageID, state, app).message;
-				await editMessage({
+				const messageObject = getMessage(editing.messageID, state, app)
+				const oldMessage = messageObject.message;
+				const editedMessage = await editMessage({
 					id: editing.messageID,
 					chatId: app.currentChatId,
 					newMessage: message,
 					oldMessage,
 				});
-				editText(editing.messageID, app.currentChatId, message, oldMessage);
-				const messageObject = getMessage(editing.messageID, state, app);
 
-				updateMessage(editing.messageID, messageObject, true);
-			} catch {
+				updateMessage({ ...editedMessage, room: app.currentChatId }, true);
+			} catch (e) {
 				setEditing({ isediting: false, messageID: null });
 				return;
 			}
@@ -272,8 +267,8 @@ const Chat = () => {
 			removeMessage(id, chatId);
 		};
 
-		const editMessageHandler = ({ id, chatId, newMessage, oldMessage }) => {
-			editText(id, chatId, newMessage, oldMessage);
+		const editMessageHandler = (messageEdited) => {
+			updateMessage({ ...messageEdited, room: app.currentChatId }, true);
 		};
 		
 		const limitMessageHandler = (data) => {
@@ -327,12 +322,14 @@ const Chat = () => {
 				>
 					{sortedMessages.map(
 						({ senderId: sender, id, message, time, status, isEdited, oldMessages, containsBadword, isRead }) => {
-							const isSender = sender.toString() === senderId.toString()
+							const isSender = sender.toString() === senderId.toString();
+							message = decryptMessage(message)
+
 							return (
 								<div
 									key={id}
 									id={!isSender && `message-${id}`}
-									className={`w-full flex text-white relative ${
+									className={`w-full flex text-white relative mb-2 ${
 										isSender ? 'justify-end' : 'justify-start'
 									}`}
 								>
@@ -402,8 +399,8 @@ const Chat = () => {
 														openPreviousMessages={openPreviousMessages}
 														oldMessages={oldMessages}
 													/>
-													<MessageSeen isRead={isRead} isSender={isSender} />
 												</div>
+												<MessageSeen isRead={isRead} isSender={isSender} />
 											</>}
 									</div>
 								</div>

--- a/client/src/components/Chat/PreviousMessages.jsx
+++ b/client/src/components/Chat/PreviousMessages.jsx
@@ -5,6 +5,11 @@ import PropTypes from 'prop-types'
 import { AiFillCaretDown } from 'react-icons/ai';
 import { BiSolidEditAlt } from 'react-icons/bi';
 
+import BadWordsNext from 'bad-words-next';
+import en from 'bad-words-next/data/en.json'
+
+import decryptMessage from 'src/lib/decryptMessage';
+
 
 const PreviousMessages = ({
 	id,
@@ -13,6 +18,8 @@ const PreviousMessages = ({
 	openPreviousEdit,
 	openPreviousMessages,
 	oldMessages }) => {
+
+	const badwords = new BadWordsNext({ data: en });
 	return (
 		<div>
 			{isEdited && (
@@ -36,11 +43,15 @@ const PreviousMessages = ({
 					<div className="flex flex-col">
 						{oldMessages !== undefined &&
 							Array.isArray(oldMessages) &&
-							oldMessages.map((message, index) => (
+							oldMessages.map((message, index) => {
+								message = badwords
+									.filter(decryptMessage(message))
+								return (
 								<span key={index} className="text-base">
 									{message}
 								</span>
-							))}
+								)
+							})}
 					</div>
 				</div>
 			)}
@@ -53,8 +64,8 @@ export default PreviousMessages
 PreviousMessages.propTypes = {
 	id: PropTypes.string.isRequired,
 	isSender: PropTypes.bool.isRequired,
-	isEdited: PropTypes.bool.isRequired,
+	isEdited: PropTypes.bool,
 	openPreviousEdit: PropTypes.func.isRequired,
-	openPreviousMessages: PropTypes.string.isRequired,
+	openPreviousMessages: PropTypes.string,
 	oldMessages: PropTypes.array
 }

--- a/client/src/context/ChatContext.jsx
+++ b/client/src/context/ChatContext.jsx
@@ -62,10 +62,10 @@ export const ChatProvider = ({ children }) => {
 	 * @param {string} id
 	 * @param {Message} message
 	 */
-	function updateMessage(id, message, isEdited) {
+	function updateMessage(message, isEdited) {
 		dispatch({
 			type: 'UPDATE_MESSAGE',
-			payload: { id, message, isEdited },
+			payload: { message, isEdited },
 		});
 	}
 
@@ -87,13 +87,6 @@ export const ChatProvider = ({ children }) => {
 		dispatch({
 			type: 'REMOVE_MESSAGE',
 			payload: { id, room: chatId },
-		});
-	}
-
-	function editText(id, chatId, newText, oldMessage) {
-		dispatch({
-			type: 'EDIT_TEXT',
-			payload: { id, room: chatId, newText, oldMessage },
 		});
 	}
 
@@ -126,7 +119,6 @@ export const ChatProvider = ({ children }) => {
 				updateMessage,
 				createChat,
 				removeMessage,
-				editText,
 				closeChat,
 				closeAllChats,
 				receiveMessage

--- a/client/src/context/reducers/chatReducer.js
+++ b/client/src/context/reducers/chatReducer.js
@@ -72,8 +72,9 @@ export default function chatReducer(state, action) {
 		}
 
 		case 'UPDATE_MESSAGE': {
-			const { id, message, isEdited } = action.payload;
+			const { message, isEdited } = action.payload;
 
+			const id = message.id;
 			if (!clonedState[message.room]) {
 				throw new Error('Room not found!');
 			}
@@ -85,6 +86,7 @@ export default function chatReducer(state, action) {
 			const room = message.room;
 			const messageId = message.id;
 			const updatedMessage = message;
+			const oldMessages = message.oldMessages;
 
 			// Assign the message to the cloned state
 			clonedState[room].messages[messageId] = updatedMessage;
@@ -94,6 +96,12 @@ export default function chatReducer(state, action) {
 				break;
 			}
 			clonedState[room].messages[messageId].isEdited = isEdited;
+
+			if (oldMessages.length === 0) {
+				break;
+			}
+
+			clonedState[room].messages[messageId].oldMessages = oldMessages 
 			break;
 		}
 

--- a/client/src/lib/browserNotification.js
+++ b/client/src/lib/browserNotification.js
@@ -1,5 +1,6 @@
 import BadWordsNext from 'bad-words-next';
 import en from 'bad-words-next/data/en.json'
+import decryptMessage from './decryptMessage';
 
 export const requestBrowserNotificationPermissions = () => {
 	if (!('Notification' in window)) {
@@ -10,7 +11,7 @@ export const requestBrowserNotificationPermissions = () => {
 };
 
 export const createBrowserNotification = (title, body) => {
-	const badwords = new BadWordsNext({ data: en })
+	const badwords = new BadWordsNext({ data: en });
 
 	if (Notification.permission === 'denied') {
 		return;
@@ -20,7 +21,7 @@ export const createBrowserNotification = (title, body) => {
 		return;
 	}
 
-	const message = badwords.filter(body);
+	const message = badwords.filter(decryptMessage(body));
 
 	new Notification(title, {
 		body: message,

--- a/client/src/lib/decryptMessage.js
+++ b/client/src/lib/decryptMessage.js
@@ -1,0 +1,10 @@
+import CryptoJS from 'crypto-js';
+
+const secretKey = import.meta.env.VITE_SECRET_KEY;
+
+export default (message) => {
+	const bytes = CryptoJS.AES.decrypt(message, secretKey);
+	const originalMessage = bytes.toString(CryptoJS.enc.Utf8);
+
+	return originalMessage;
+};

--- a/server/package.json
+++ b/server/package.json
@@ -20,6 +20,7 @@
   "license": "ISC",
   "dependencies": {
     "cors": "^2.8.5",
+    "crypto-js": "^4.1.1",
     "dotenv": "^16.3.1",
     "express": "^4.18.2",
     "express-rate-limit": "^7.1.1",

--- a/server/sockets/editMessage.js
+++ b/server/sockets/editMessage.js
@@ -17,18 +17,13 @@ module.exports = (socket) => {
         return;
       }
 
-      const messageEditted = await editMessage(chatId, {
+      const messageEdited = await editMessage(chatId, {
         id: messageId,
         message: newMessage,
         oldMessage,
       });
-      socket.broadcast.to(chatId).emit(NEW_EVENT_EDIT_MESSAGE, {
-        id: messageId,
-        chatId,
-        newMessage,
-        oldMessage,
-      });
-      messageWasEditedSuccessfully(messageEditted);
+      socket.broadcast.to(chatId).emit(NEW_EVENT_EDIT_MESSAGE, messageEdited);
+      messageWasEditedSuccessfully(messageEdited);
     }
   );
 };

--- a/server/utils/lib.js
+++ b/server/utils/lib.js
@@ -1,8 +1,12 @@
 const { Socket } = require('socket.io');
 const mongoose = require('mongoose');
+const CryptoJS = require('crypto-js');
+
 const ActiveUser = require('../models/ActiveUserModel');
 const Chat = require('../models/ChatModel');
 const Message = require('../models/MessageModel');
+
+const secretKey = process.env.SECRET_KEY;
 
 /**
  * @typedef {{
@@ -296,6 +300,8 @@ async function addMessage(
 ) {
   const sender = getActiveUser(senderId);
 
+  message = CryptoJS.AES.encrypt(message, secretKey).toString();
+
   if (!sender) {
     return null;
   }
@@ -360,6 +366,8 @@ async function editMessage(chatId, { id, message, oldMessage }) {
     return false;
   }
 
+  message = CryptoJS.AES.encrypt(message, secretKey).toString();
+
   try {
     await Message.findOneAndUpdate(
       { _id: id },
@@ -376,11 +384,10 @@ async function editMessage(chatId, { id, message, oldMessage }) {
       chats[chatId].messages[id].oldMessages = [];
     }
     chats[chatId].messages[id].oldMessages.push(oldMessage);
+    return chats[chatId].messages[id];
   } catch {
     return false;
   }
-
-  return true;
 }
 
 async function seenMessage(chatId, messageId) {

